### PR TITLE
Trace improvements

### PR DESCRIPTION
**Pull request by manasdk**
* Include a `caused_by` field in each component
* It is now possible to follow the causality chain in the CLI i.e. instead seeing a giant list of associated components of a Trace we can now see a causality chain.

### show executions
```
$ st2 trace get 56662b4032ed355770be9f73 --show-executions
id: 56662b4032ed355770be9f73
trace_tag: trigger_instance-56662b4032ed355770be9f72
start_timestamp: 2015-12-08T00:58:40.434009Z
+--------------------------+-----------+-----------------------+
| id                       | type      | updated_at            |
+--------------------------+-----------+-----------------------+
| 56662b4032ed355770be9f75 | execution | 2015-12-08T00:58:40.5 |
|                          |           | 35939Z                |
+--------------------------+-----------+-----------------------+
```

### show rules
```
$ st2 trace get 56662b4032ed355770be9f73 --show-rules
id: 56662b4032ed355770be9f73
trace_tag: trigger_instance-56662b4032ed355770be9f72
start_timestamp: 2015-12-08T00:58:40.434009Z
+--------------------------+------+-----------------------+
| id                       | type | updated_at            |
+--------------------------+------+-----------------------+
| 5666295032ed355785437236 | rule | 2015-12-08T00:58:40.4 |
|                          |      | 70995Z                |
+--------------------------+------+-----------------------+
```

### show trigger instances
```
$ st2 trace get 56662b4032ed355770be9f73 --show-trigger-instances
id: 56662b4032ed355770be9f73
trace_tag: trigger_instance-56662b4032ed355770be9f72
start_timestamp: 2015-12-08T00:58:40.434009Z
+--------------------------+------------------+-----------------------+
| id                       | type             | updated_at            |
+--------------------------+------------------+-----------------------+
| 56662b4032ed355770be9f72 | trigger-instance | 2015-12-08T00:58:40.4 |
|                          |                  | 34504Z                |
| 56662b4032ed355770be9f76 | trigger-instance | 2015-12-08T00:58:40.7 |
|                          |                  | 23010Z                |
+--------------------------+------------------+-----------------------+
```

### causation chains (starting from ActionExecution)
```
$ st2 trace get 56662b4032ed355770be9f73 -e 56662b4032ed355770be9f75
id: 56662b4032ed355770be9f73
trace_tag: trigger_instance-56662b4032ed355770be9f72
start_timestamp: 2015-12-08T00:58:40.434009Z
+--------------------------+------------------+-----------------------+
| id                       | type             | updated_at            |
+--------------------------+------------------+-----------------------+
| 56662b4032ed355770be9f72 | trigger-instance | 2015-12-08T00:58:40.4 |
|                          |                  | 34504Z                |
| 5666295032ed355785437236 | rule             | 2015-12-08T00:58:40.4 |
|                          |                  | 70995Z                |
| 56662b4032ed355770be9f75 | execution        | 2015-12-08T00:58:40.5 |
|                          |                  | 35939Z                |
+--------------------------+------------------+-----------------------+
```

### CODE REVIEW

**Comment by Kami** [comment]
Cool.

It looks like this also ties into the whole rule enforcement stuff, right? Do we still need that PR / feature after this change?

And if we do, can that PR just utilize tracing to retrieve this information and display it in a "nicer" manner?

**Comment by manasdk** [comment]
> It looks like this also ties into the whole rule enforcement stuff, right? Do we still need that PR / feature after this change?

Yes. Trace needs to use a RuleEnforcement object rather than a rule object.

> And if we do, can that PR just utilize tracing to retrieve this information and display it in a "nicer" manner?

The requirements are different for these features. 

`Trace` - what is the full context of "activity" that happened.

`RuleEnforcement` - was this rule enforced or not. This itself would be part of the Trace.

File: st2common/st2common/models/api/trace.py [hunk 1]
```diff
@@ -30,6 +30,20 @@
             'description': 'The start time when the action is executed.',
             'type': 'string',
             'pattern': isotime.ISO8601_UTC_REGEX
+        },
+        'causal_component': {
```
Comment (lakshmi-kannan) [inline st2common/st2common/models/api/trace.py]: It's a mouthful. how about caused_by?

**Comment by lakshmi-kannan** [comment]
Is the end goal to show some relationship between items in CLI and UI?

**Pull request merged**